File: node.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
	int data;
	struct Node * left;
	struct Node * right;
} node ;

int get_data(node* n) {
	return n->data;
}

void assign_data(node* n, int data) {
	n->data = data;
}

int lookup(node* n, int target) {
	if (n == NULL) {
		return 0;
	} else {
		if (target == n->data) {
			return 1;
		} else {
			if (target < n->data) {
				return lookup(n->left, target);
			}
			else {
				return lookup(n->right, target);
			}
		}
	}
}

node* new_node(int data) {
	node* n = malloc(sizeof(node));
	n->data = data;
	n->left = NULL;
	n->right = NULL;
	return n;
}

node* insert(node* n, int data) {
	if (n == NULL) {
		return new_node(data);
	} else {
		if (data <= n->data) {
			n->left = insert(n->left, data);
		} else {
			n->right = insert(n->right, data);
		}
	}
	return n;
}

void free_nodes(node* n) {
	if (n == NULL) {
		return;
	}

	free_nodes(n->left);
	free_nodes(n->right);
	free(n);
}

int is_leaf(node* n) {
	return n->left == NULL && n->right == NULL;
}

void print_nodes_with_order(node* n, int order) {
	// preorder = 0, inorder = 1, postorder = 2
	// otherwise it won't print :P
	if (order == 0) {
		printf("%d\n", n->data);
	}

	if (n->left != NULL) {
		print_nodes_with_order(n->left, order);
	}

	if (order == 1) {
		printf("%d\n", n->data);
	}

	if (n->right != NULL) {
		print_nodes_with_order(n->right, order);
	}

	if (order == 2) {
		printf("%d\n", n->data);
	}
}

int count_nodes(node* n) {
	if (n == NULL) {
		return 0;
	}
	
	int count = 1;

	if (n->left != NULL) {
		count += count_nodes(n->left);
	}
	if (n->right != NULL) {
		count += count_nodes(n->right);
	}
	if (n == NULL) {
		return 0;
	}

	return count;
}

File: node.py
#! /usr/bin/env python3

import sys


class Node(object):
    def __init__(self, data, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right


def insert_node(node, data):
    if not(node):
        return Node(data)

    else:
        if data <= node.data:
            node.left = insert_node(node.left, data)

        else:
            node.right = insert_node(node.right, data)

    return node


def lookup(node, target):
    if not(node):
        return False

    else:
        if target == node.data:
            return True
        elif target < node.data:
            return lookup(node.left, target)
        else:
            return lookup(node.right, target)


tree = Node(int(sys.argv[1]))

for val in sys.argv[2:]:
    # tree = insert_node(tree, int(val))
    insert_node(tree, int(val))

for val in sys.argv[1:]:
    print(lookup(tree, int(val)))
